database/caching improvements

muXxer opened the pull request:
Fixes #914

muXxer committed:
Remove badger

muXxer committed:
Optimize pebble settings to prevent read amplified triggered compactions

muXxer committed:
Improve caching of messages

luca-moser reviewed (approved):
I'm for merging this even though I can't really say what effects the new changes to the Pebble settings will have.

muXxer commented:
Just for the reference here as a short explanation.

I tried to stick to the pebble defaults in most of the cases, and added all the comments from pebble so we can directly see what these parameters are for.

The moment we added the "compaction" stats to prometheus, we saw that after some time the node starts a compaction run, which somehow never stops. This leads to massive performance losses, since the disc IO and CPU usage is almost at maximum all the time. You can also see that described in the issue #914.

I was wondering what causes this, since we were using default settings. It seems that the read amplified triggered compactions cause this. Maybe a compaction also causes "read access" which then causes an compaction again, so we loop infinitely. After changing the values that cause "read amplified triggered" compactions to different values, the infinite compaction runs stopped. Now we see only small compaction runs, which seem to be triggered by "write amplification" only. I'm not a database expert, but I guess this is enough for our use case. We shouldn't encounter write stalls, because we still have write amplification triggered compactions.
